### PR TITLE
[#208] drop-databases with --sudo option throws an error 

### PR DESCRIPTION
**Pull request by brunosedler**
closes https://github.com/makandra/geordi/issues/208

### CODE REVIEW

File: lib/geordi/db_cleaner.rb [hunk 1]
```diff
@@ -126,7 +132,7 @@ def decide_mysql_command(extra_flags)
             cmd = 'mysql -uroot'
             cmd << " #{extra_flags}" unless extra_flags.nil?
             unless File.exist? File.join(Dir.home, '.my.cnf')
-              puts "Please enter your MySQL/MariaDB password for account 'root'."
+              puts "> Please enter your MySQL/MariaDB password for account 'root'."
```
Comment (codener) [inline lib/geordi/db_cleaner.rb:135]:
```suggestion
              Interaction.note "Please enter your MySQL/MariaDB password for account 'root'."
```

File: lib/geordi/db_cleaner.rb [hunk 2]
```diff
@@ -165,17 +171,27 @@ def list_all_dbs(dbtype)
       else
         list_all_mysql_dbs
       end
+    rescue DatabaseError
+      Interaction.fail 'Connection to database could not be established. Try running the command again with --sudo'
```
Comment (codener) [inline lib/geordi/db_cleaner.rb:175]:
```suggestion
      Interaction.fail 'Connection to database could not be established. Try running again with --sudo.'
```

File: lib/geordi/db_cleaner.rb [hunk 3]
```diff
@@ -165,17 +171,27 @@ def list_all_dbs(dbtype)
...
     end
 
     def list_all_postgres_dbs
-      `#{@postgres_command} -t -A -c 'SELECT DATNAME FROM pg_database WHERE datistemplate = false'`.split
+      output, error, _status = Open3.capture3("#{@postgres_command} -t -A -c 'SELECT DATNAME FROM pg_database WHERE datistemplate = false'")
+
+      raise DatabaseError unless error.empty?
```
Comment (codener) [inline lib/geordi/db_cleaner.rb:181]: Usually, the status is an even better means of recognizing an error. You can even ask `status.success?`. Some commands write to stderr even if overall successful.

File: lib/geordi/db_cleaner.rb [hunk 4]
```diff
@@ -186,7 +202,7 @@ def clean_mysql
       return if deletable_dbs.nil?
       deletable_dbs.each do |db|
         if @mysql_command.include? '-p'
-          puts "Please enter your MySQL/MariaDB account 'root' for: DROP DATABASE #{db}"
+          puts "> Please enter your MySQL/MariaDB account 'root' for: DROP DATABASE #{db}"
```
Comment (codener) [inline lib/geordi/db_cleaner.rb:205]:
```suggestion
          Interaction.note "Please enter your MySQL/MariaDB root password (for running DROP DATABASE #{db}").
```

File: lib/geordi/db_cleaner.rb [hunk 5]
```diff
@@ -3,24 +3,30 @@
...
+  DatabaseError = Class.new(StandardError)
+
   class DBCleaner
 
     def initialize(extra_flags, sudo: false)
       @sudo = sudo
 
       if @sudo
-        puts 'Please enter your sudo password if asked, for db operations as system users'
+        puts '> Please enter your sudo password if asked, for db operations as system users'
```
Comment (codener) [inline lib/geordi/db_cleaner.rb:15]:
```suggestion
        Interaction.note 'Please enter your sudo password when asked.'
```

File: lib/geordi/db_cleaner.rb [hunk 6]
```diff
@@ -165,17 +171,27 @@ def list_all_dbs(dbtype)
...
+
+      raise DatabaseError unless error.empty?
+
+      output.split
     end
 
     def list_all_mysql_dbs
       if @mysql_command.include? '-p'
-        puts "Please enter your MySQL/MariaDB account 'root' for: list all databases"
+        puts "> Please enter your MySQL/MariaDB account 'root' for: list all databases"
```
Comment (codener) [inline lib/geordi/db_cleaner.rb:188]:
```suggestion
        Interaction.note "Please enter your MySQL/MariaDB root password (for listing all databases)."
```

File: spec/db_cleaner_spec.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    before do
+      stub_const('ENV', ENV.to_hash.merge('XDG_CONFIG_HOME' => dir))
+    end
+
+    after do
+      file_dir = File.join(dir, '/.config')
+      FileUtils.rm_rf(file_dir)
+    end
+
+    it 'moves the legacy allowlists from ~/.config/geordi/whitelists to ~/.config/geordi/allowlists' do
```
Comment (codener) [inline spec/db_cleaner_spec.rb:16]: Normally, we would have written these specs as Cucumber Aruba features. There is really nice command line support, also for working with files. But it's really nice you got this working in RSpec as well. Let's keep it.

File: spec/db_cleaner_spec.rb [hunk 8]
```diff
@@ -0,0 +1,67 @@
+require 'fileutils'
+
+RSpec.describe Geordi::DBCleaner do
+  describe '#new' do
+    let(:dir) { File.join(Dir.pwd, 'spec') }
```
Comment (codener) [inline spec/db_cleaner_spec.rb:5]: Please write to ./tmp in tests. Maybe tmp is already excluded in the gitignore file?

**Review by codener — CHANGES_REQUESTED** [review]
Thanks for the update, nice implementation.

I have a few remarks.

File: spec/db_cleaner_spec.rb (hunk 8 above)
Comment (brunosedler) [reply spec/db_cleaner_spec.rb]: Good point, and it is indeed already excluded :+1:

File: spec/db_cleaner_spec.rb (hunk 7 above)
Comment (brunosedler) [reply spec/db_cleaner_spec.rb:16]: I prefer writing rspec to cucumber :sweat_smile: 
For testing commands as a whole cucumber certainly makes sense, but this seemed more like something I should test in a unit test, as it diverges from the "happy path"  imo.

**Review by codener — APPROVED** [review]
Fine!

**Pull request merged**